src/opencl_memory.c

```diff
@@ -0,0 +1,248 @@
+	if (!platform_ids) {
+		printf("Allocation of platform_ids failed\n");
+		return FAILURE;
+	}
+
+	device_ids = malloc(sizeof(cl_device_id) * (ctx->device_ix + 1));
+	if (!device_ids) {
+		printf("Allocation of device_ids failed\n");
+		ret = FAILURE;
+		goto free_platform_ids;
+	}
+
+	error = clGetPlatformIDs((ctx->platform_ix + 1), platform_ids, &num_platforms);
+	if (error) {
+		printf("clGetPlatformIDs returned %d\n", error);
+		ret = FAILURE;
+		goto free_device_ids;
+	}
+
+	if (num_platforms < ctx->platform_ix)
+	{
+		printf("platform_id (%d) is not in the range of num_platforms (%d)\n",
+				ctx->platform_ix, num_platforms);
+		ret = FAILURE;
+		goto free_device_ids;
+	}
+
+	error = clGetDeviceIDs(platform_ids[ctx->platform_ix], CL_DEVICE_TYPE_ALL, (ctx->device_ix + 1), device_ids, &num_devices);
+	if (error) {
+		printf("clGetDeviceIDs returned %d\n", error);
+		ret = FAILURE;
+		goto free_device_ids;
+	}
+
+	if (num_devices < ctx->device_ix)
+	{
+		printf("device_id (%d) is not in the range of num_devices (%d)\n",
+				ctx->device_ix, num_devices);
+		ret = FAILURE;
+		goto free_device_ids;
+	}
+
+	error = clGetDeviceInfo(device_ids[ctx->device_ix], CL_DEVICE_SVM_CAPABILITIES, sizeof(cl_device_svm_capabilities), &caps, 0);
+	if (error) {
+		printf("clGetDeviceInfo returned %d\n", error);
+		ret = FAILURE;
+		goto free_device_ids;
+	}
+
+	if (!(caps & CL_DEVICE_SVM_FINE_GRAIN_SYSTEM))
+	{
+		printf("SVM needed caps are not supported\n");
+		ret = FAILURE;
+		goto free_device_ids;
+	}
+
+	ctx->cl_context = clCreateContext(NULL, 1, &device_ids[ctx->device_ix], NULL, NULL, &error);
+	if (error) {
+		printf("clCreateContext returned %d\n", error);
+		ret = FAILURE;
+		goto free_device_ids;
+	}
+
+	ctx->command_queue = clCreateCommandQueueWithProperties(ctx->cl_context, device_ids[ctx->device_ix], NULL, &error);
+	if (error) {
+		printf("clCreateCommandQueueWithProperties failed with ret=%d\n", error);
+		clReleaseContext(ctx->cl_context);
+		ret = FAILURE;
+	}
+
+free_device_ids:
+	free(device_ids);
+free_platform_ids:
+	free(platform_ids);
+
+	return ret;
+}
+
+static void free_gpu(struct opencl_memory_ctx *ctx)
+{
+	printf("destroying current OpenCL ctx\n");
+	clReleaseCommandQueue(ctx->command_queue);
+	clReleaseContext(ctx->cl_context);
+}
+
+int opencl_memory_init(struct memory_ctx *ctx)
+{
+	struct opencl_memory_ctx *opencl_ctx = container_of(ctx, struct opencl_memory_ctx, base);
+	int return_value = 0;
+
+	return_value = init_gpu(opencl_ctx);
+	if (return_value) {
+		fprintf(stderr, "Couldn't init GPU context: %d\n", return_value);
+		return FAILURE;
+	}
+
+	return SUCCESS;
+}
+
+int opencl_memory_destroy(struct memory_ctx *ctx) {
+	struct opencl_memory_ctx *opencl_ctx = container_of(ctx, struct opencl_memory_ctx, base);
+
+	free_gpu(opencl_ctx);
+	free(opencl_ctx);
+	return SUCCESS;
+}
+
+void *touch_gpu_pages(void *ctx_param) {
+	struct buffer_ctx *ctx = (struct buffer_ctx *)ctx_param;
+	int ret;
```

Comment:
please add `// cppcheck-suppress variableScope` next to `int ret;` or move it into the loop scope.